ledger/src/blockstore.rs

```diff
@@ -2235,13 +2235,9 @@ impl Blockstore {
         if let Some(highest_primary_index_slot) = *w_highest_primary_index_slot {
             if oldest_slot > highest_primary_index_slot {
                 *w_highest_primary_index_slot = None;
-                self.transaction_status_index_cf.delete(0)?;
-                self.transaction_status_index_cf.delete(1)?;
+                self.db.set_clean_slot_0(true);
             }
         }
-        if w_highest_primary_index_slot.is_none() {
-            self.db.set_clean_slot_0(true);
-        }
```

Comment:
I'm wondering about this change. Is L2237 the only place we set `w_highest_primary_index_slot = None` anew?

Reply:
I believe the locations that modify `highest_primary_index_slot` are:
- `Blockstore::maybe_cleanup_highest_primary_index_slot()` (this function)
- `Blockstore::update_highest_primary_index_slot()`
- `Blockstore::set_highest_primary_index_slot()`

`maybe_cleanup_highest_primary_index_slot()` gets called by `set_max_expired_slot()`, which is called from the newly renamed `BlockstoreCleanupService`.

`update_highest_primary_index_slot()` gets called once and only once right at the end of `Blockstore::do_open()`

`set_highest_primary_index_slot()` can get called directly by `Blockstore::purge_special_columns_exact()`; however, recent optimizations here mean the function bails out early when the special columns are not being filled.

Reply:
That being said, I don't think we want to delete the transaction_status_index entries right now. We need them present in v1.16/v1.17 branches, so this scenario could result in an error if we leave it:
- Operator running RPC node (special columns populated) with v1.17
- Operator upgrades to master (or let's pretend v1.18)
- After running for a bit, the node will realize that is has purged all data from when it was still updating the  TransactionStatusIndex entries. Thus, `maybe_cleanup_highest_primary_index_slot()` purges the entries
- The operator attempts to run v1.17 solana-ledger-tool command on their node; it fails because v1.17 expects the TransactionStatusIndex entries to be present but they are not

I believe this is the sequence Brooks hit, but let me double check. As mentioned in the PR issue, there has been some churn on this so I would like to fully understand exactly what happened

Reply:
To be clear, I was only questioning moving `self.db.set_clean_slot_0(true)` into the `if let-if` case. It took a bit, but I've convinced myself it is sound.

- `update_highest_primary_index_slot()` calls `self.db.set_clean_slot_0(true)` itself if the entries don't exist, or are freshly added (max slot = 0), ie. if it's decided to leave `highest_primary_index_slot` as None
- `set_highest_primary_index_slot()` only ever sets `highest_primary_index_slot` to Some